// @suejon/youtube-subtitles/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// index.ts
var youtube_subtitles_exports = {};
__export(youtube_subtitles_exports, {
  get_available_languages: () => get_available_languages,
  get_subtitles_for_video: () => get_subtitles_for_video
});
module.exports = __toCommonJS(youtube_subtitles_exports);

// constants.ts
var WATCH_URL = "https://www.youtube.com/watch?v=";

// utils.ts
function isTranscripsMeta(value) {
  return "translationLanguages" in value && "captionTracks" in value;
}
async function fetch_transcript(url) {
  const timed_text = await fetch(url, {
    headers: { "Accept-Language": "en-us" }
  });
  const html = await timed_text.text();
  try {
    const parser = new DOMParser();
    const xml_doc = parser.parseFromString(html, "application/xml");
    const captions = [];
    const root = xml_doc.childNodes[0];
    if (!root) {
      throw new Error("no root node found");
    }
    for (let i = 0; i < root.childNodes.length; i++) {
      const element = root.childNodes[i];
      if (!element) {
        continue;
      }
      const text = element.textContent;
      const start = element.getAttribute("start");
      const dur = element.getAttribute("dur");
      if (!text || !start || !dur) {
        continue;
      }
      captions.push({
        text: element.textContent?.replace(/<[^>]*>/, "") || "",
        start: parseInt(start),
        duration: parseInt(dur) ?? 0
      });
    }
    return captions;
  } catch (error) {
    console.error("error: ", error);
  }
}
async function list_transcripts(video_id) {
  console.log("fetching transcripts for video id: ", video_id);
  const data = await fetch(WATCH_URL + video_id, {
    headers: { "Accept-Language": "en-us" }
  });
  const html = await data.text();
  const captions = extract_json(html);
  if (!isTranscripsMeta(captions)) {
    throw new Error("error extracting captions");
  }
  return captions;
}
function extract_json(html) {
  const split_html = html.split(/"captions":/);
  if (split_html.length === 1) {
    throw new Error("captions not found");
  }
  try {
    const captions_text = split_html[1]?.split(',"videoDetails"')[0]?.replace("\n", "");
    if (!captions_text) {
      throw new Error("captions not found");
    }
    const captions_json = JSON.parse(captions_text).playerCaptionsTracklistRenderer;
    if (!captions_json) {
      throw new Error("captions not found");
    }
    return captions_json;
  } catch (error) {
    throw new Error("captions not found");
  }
}

// index.ts
async function get_subtitles_for_video(video_id, language = "en") {
  const transcriptsMeta = await list_transcripts(video_id);
  const transcriptLang = transcriptsMeta.translationLanguages.filter(
    (lang) => lang.languageCode === language
  );
  if (transcriptLang.length === 0) {
    throw new Error("No transcript found for language: " + language);
  }
  let url = transcriptsMeta.captionTracks[0]?.baseUrl;
  if (url === void 0) {
    throw new Error("No transcript URL found");
  }
  if (language !== "en") {
    url = url + "&tlang=" + language;
  }
  return fetch_transcript(url);
}
async function get_available_languages(video_id) {
  const transcriptsMeta = await list_transcripts(video_id);
  return transcriptsMeta.translationLanguages.map((x) => x.languageCode);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  get_available_languages,
  get_subtitles_for_video
});
